packages/vscode-extension/src/commonlib/vscodeAzureSubscriptionProvider.ts

```diff
@@ -0,0 +1,263 @@
+        };
+      },
+    };
+
+    const configuredAzureEnv = getConfiguredAzureEnv();
+    const endpoint = configuredAzureEnv.resourceManagerEndpointUrl;
+
+    return {
+      client: new armSubs.SubscriptionClient(credential, { endpoint }),
+      credential: credential,
+      authentication: {
+        getSession: () => session,
+      },
+    };
+  }
+
+  /**
+   * Gets a list of tenants available to the user.
+   *
+   * @returns A list of tenants.
+   */
+  public async getTenants(): Promise<TenantIdDescription[]> {
+    const { client } = await this.getSubscriptionClient(undefined, AzureScopes);
+
+    const results: TenantIdDescription[] = [];
+
+    for await (const tenant of client.tenants.list()) {
+      results.push(tenant);
+    }
+
+    return results;
+  }
+
+  /**
+   * Gets a list of Azure subscriptions available to the user.
+   */
+  public async getSubscriptions(): Promise<AzureSubscription[]> {
+    const results: AzureSubscription[] = [];
+
+    for (const tenant of await this.getTenants()) {
+      try {
+        // Get the list of tenants
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        const tenantId = tenant.tenantId!;
+
+        // For each tenant, get the list of subscriptions
+        results.push(...(await this.getSubscriptionsForTenant(tenantId)));
+      } catch (e) {}
+    }
+    const sortSubscriptions = (subscriptions: AzureSubscription[]): AzureSubscription[] =>
+      subscriptions.sort((a, b) => a.name.localeCompare(b.name));
+    return sortSubscriptions(results);
+  }
+
+  /**
+   * Gets the subscriptions for a given tenant.
+   *
+   * @param tenantId The tenant ID to get subscriptions for.
+   *
+   * @returns The list of subscriptions for the tenant.
+   */
+  private async getSubscriptionsForTenant(tenantId: string): Promise<AzureSubscription[]> {
+    const { client, credential, authentication } = await this.getSubscriptionClient(
+      tenantId,
+      AzureScopes
+    );
+    const environment = getConfiguredAzureEnv();
+
+    const subscriptions: AzureSubscription[] = [];
+
+    for await (const subscription of client.subscriptions.list()) {
+      subscriptions.push({
+        authentication: authentication,
+        environment: environment,
+        credential: credential,
+        isCustomCloud: environment.isCustomCloud,
+        /* eslint-disable @typescript-eslint/no-non-null-assertion */
+        name: subscription.displayName!,
+        subscriptionId: subscription.subscriptionId!,
+        /* eslint-enable @typescript-eslint/no-non-null-assertion */
+        tenantId: tenantId,
+      });
+    }
+
+    return subscriptions;
+  }
+}
+
+export async function getSessionFromVSCode(
+  scopes?: string | string[],
+  tenantId?: string,
+  options?: vscode.AuthenticationGetSessionOptions
+): Promise<vscode.AuthenticationSession | undefined> {
+  return await vscode.authentication.getSession(Microsoft, getScopes(scopes, tenantId), options);
```

Comment:
What's the minimal vsc version supporting this new auth flow? Do we need to update the `engines -> vscode` in ext's *package.json*?